arch/risc-v/src/esp32c3/esp32c3_twai.c

```diff
@@ -0,0 +1,1265 @@
+  for (tseg = (0 + 0 + 2) * 2;
+       tseg <= (timing->tseg2_max + timing->tseg1_max + 2) * 2 + 1; tseg++)
+    {
+      brp = clock / ((1 + tseg / 2) * rate) + tseg % 2;
+      if (brp == 0 || brp > 64)
+        {
+          continue;
+        }
+
+      error = rate - clock / (brp * (1 + tseg / 2));
+      if (error < 0)
+        {
+          error = -error;
+        }
+
+      if (error <= best_error)
+        {
+          best_error = error;
+          best_tseg = tseg / 2;
+          best_brp = brp;
+          best_rate = clock / (brp * (1 + tseg / 2));
+        }
+    }
+
+  if (best_error && (rate / best_error < 10))
+    {
+      caninfo("baud rate %d is not possible with %d Hz clock\n",
+              rate, clock);
+      caninfo("%d bps. brp=%d, best_tseg=%d, tseg1=%d, tseg2=%d\n",
+              best_rate, best_brp, best_tseg, tseg1, tseg2);
+      return -EINVAL;
+    }
+
+  tseg2 = best_tseg - (sampl_pt * (best_tseg + 1)) / 100 + 1;
+  if (tseg2 < 0)
+    {
+      tseg2 = 0;
+    }
+
+  if (tseg2 > timing->tseg2_max)
+    {
+      tseg2 = timing->tseg2_max;
+    }
+
+  tseg1 = best_tseg - tseg2;
+  if (tseg1 > timing->tseg1_max)
+    {
+      tseg1 = timing->tseg1_max;
+      tseg2 = best_tseg - tseg1;
+    }
+
+  caninfo("TS1: %d TS2: %d BRP: %d\n", tseg1, tseg2, best_brp);
+
+  /* Configure bit timing */
+
+    timing0 = (best_brp - 1) / 2;
+    timing0 |= (sjw - 1) << TWAI_SYNC_JUMP_WIDTH_S;
+
+    timing1 = tseg1 - 1;
+    timing1 |= (tseg2 - 1) << TWAI_TIME_SEG2_S;
+
+  #ifdef CONFIG_ESP32C3_TWAI0_SAM
+  /* The bus is sampled 3 times (recommended for low to medium speed buses
+   * to spikes on the bus-line).
+   */
+
+      timing1 |= CONFIG_ESP32C3_TWAI0_SAM << TWAI_TIME_SAMP_S;
+  #endif
+
+    twai_putreg(TWAI_BUS_TIMING_0_REG, timing0);
+    twai_putreg(TWAI_BUS_TIMING_1_REG, timing1);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp32c3_twaiinitialize
+ *
+ * Description:
+ *   Initialize the selected TWAI port
+ *
+ * Input Parameters:
+ *   Port number (for hardware that has multiple TWAI interfaces)
+ *
+ * Returned Value:
+ *   Valid TWAI device structure reference on success; a NULL on failure
+ *
+ ****************************************************************************/
+
+struct can_dev_s *esp32c3_twaiinitialize(int port)
+{
+  struct can_dev_s *twaidev;
+  irqstate_t flags;
+
+  caninfo("TWAI%" PRIu8 "\n",  port);
+
+  flags = enter_critical_section();
+
+#ifdef CONFIG_ESP32C3_TWAI0
+  if (port == 0)
+    {
+      /* Enable power to the TWAI module and
+       * Enable clocking to the TWAI module
+       */
+
+      modifyreg32(SYSTEM_PERIP_RST_EN0_REG, 0, SYSTEM_TWAI_RST_M);
+      modifyreg32(SYSTEM_PERIP_CLK_EN0_REG, SYSTEM_TWAI_CLK_EN_M, 0);
+
+      usleep(1);
```

Comment:
usleep depends on system tick. I recommend to use `up_udelay()` if more precise delay is needed

Reply:
This is question if the delay is really needed and for sure timing can be arbitrary, problem can be if the driver init can be called before system clock tick is available. But actual setup works.

Can somebody from Espressif comment (@acassis ) what is the minimal correct sequence to be sure that clocks are available and reset is really propagated and applied to the peripheral?

The clock enable has been added because some global location disables the clocks the fisrt, probably to lower power consumption. That is a right but what is the best switching sequence and what way to be sure that reset really clears the peripheral?

It seems that xtensa esp32 copy of the driver does not enable clocks but if the same update else were is done than it needs update as well.